redwood-treehouse-lazylayout-view/src/main/kotlin/app/cash/redwood/treehouse/lazylayout/view/ViewLazyColumn.kt

```diff
@@ -15,118 +15,156 @@
-    private val treehouseApp: TreehouseApp<A>,
-    private val widgetSystem: WidgetSystem,
-    private val contentHeight: Int,
-  ) : ListAdapter<LazyContentItem, ViewHolder>(LazyContentItemDiffCallback) {
-    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder {
-      val container = FrameLayout(parent.context).apply {
-        layoutParams = ViewGroup.LayoutParams(MATCH_PARENT, contentHeight)
+  override fun intervals(intervals: List<LazyListIntervalContent>) {
+    // TODO Don't hardcode pageSizes
+    // TODO Enable placeholder support
+    // TODO Set a maxSize so we don't keep _too_ many views in memory
+    val pager = Pager(PagingConfig(pageSize = 30, initialLoadSize = 30, enablePlaceholders = false)) {
+      ItemPagingSource(treehouseApp, intervals)
+    }
+    scope.launch {
+      pager.flow.collectLatest { pagingData ->
+        adapter.submitData(pagingData)
       }
-      return ViewHolder(container, widgetSystem)
     }
+  }
 
-    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
-      val itemContent = currentList[position]
-      val itemContentSource = TreehouseContentSource<A> {
-        itemContent.item.get(itemContent.index)
-      }
+  private class ItemPagingSource<A : AppService>(
+    private val treehouseApp: TreehouseApp<A>,
+    private val intervals: List<LazyListIntervalContent>,
+  ) : PagingSource<Int, Content>() {
 
-      holder.widgetContentBinding?.close()
-      holder.widgetContentBinding = itemContentSource.bindWhenReady(
-        view = holder.treehouseWidgetView,
-        app = treehouseApp,
+    override suspend fun load(
+      params: LoadParams<Int>,
+    ): LoadResult<Int, Content> {
+      val key = params.key ?: 0
+      val count = intervals.sumOf(LazyListIntervalContent::count)
+      val limit = when (params) {
+        is LoadParams.Prepend<*> -> minOf(key, params.loadSize)
+        else -> params.loadSize
+      }.coerceAtMost(count)
+      val offset = when (params) {
+        is LoadParams.Prepend<*> -> maxOf(0, key - params.loadSize)
+        is LoadParams.Append<*> -> key
+        is LoadParams.Refresh<*> -> if (key >= count) maxOf(0, count - params.loadSize) else key
+      }
+      val nextPosToLoad = offset + limit
+      val loadResult = LoadResult.Page(
+        data = List(limit) { index ->
+          val itemContentSource = TreehouseContentSource<A> {
+            var interval = IndexedValue(index + offset, intervals.first())
+            for (nextInterval in intervals.drop(1)) {
+              if (interval.index < interval.value.count) break
+              interval = IndexedValue(interval.index - interval.value.count, nextInterval)
+            }
+            interval.value.itemProvider.get(interval.index)
+          }
+          treehouseApp.createContent(itemContentSource).apply {
+            // TODO Pass in actual HostConfiguration
+            preload(HostConfiguration(darkMode = true))
```

Comment:
TODO: create a public API that accepts an Android Context and returns the corresponding HostConfiguration